DD-1229 schema tests

jo-pol opened the pull request:
Fixes DD-1229 schema tests

# Description of changes

* Fixed
  *  language encoding ISO639 mix-up: 1+2 in DDMv1, 2+3 in DDMv2
  * 'v1/provenance.xsd` referred to `v1/ddm.xsd`
  * `extern/xml.xsd` only loaded without the line
     `<!DOCTYPE xs:schema PUBLIC "-//W3C//DTD XMLSCHEMA 200102//EN" "XMLSchema.dtd" >`
* The class `ValidXmlsTest` shows that the all XML files in the test resources are valid. Used applicable examples from https://github.com/DANS-KNAW/easy-schema-examples. No need to change the code when adding/removing examples.
* The class `DdmVersionChangesTest` documents the effect of the differences between V1 and V2 of the DDM schemas.
* `src/test/resources/README.md` adapted from https://github.com/DANS-KNAW/easy-schema-examples

# How to test

# Related PRs

* https://github.com/DANS-KNAW/dd-dans-sword2-examples/pull/8

# Notify

@DANS-KNAW/dataversedans

jo-pol committed:
initial tests for ddm

jo-pol committed:
methodResource

jo-pol committed:
parameterize over all valid resources

jo-pol committed:
document differences of ddm versions

jo-pol committed:
everything applicable from easy-schema-examples

jo-pol committed:
license headers

jo-pol committed:
allow examples for other XSDs without changing code, test all schemas

jo-pol committed:
fixed schemas

jo-pol committed:
less loading of ddm schemas, language encoding schemas

jo-pol committed:
renames

janvanmansum commented on lib/src/test/resources/README.MD:
```diff
@@ -0,0 +1,88 @@
+| `ddm`        | `metadata/dataset.xml`                   |
+| `files`      | `metadata/files.xml`                     |
+| `agreements` | `metadata/depositor-info/agreements.xml` |
+
```
Should reference v1.0.0 not version 0.0.0

janvanmansum commented on lib/src/test/resources/README.MD:
```diff
@@ -0,0 +1,88 @@
+Example Resources
```
Move this documentation to the docs site. Maybe just create an EXAMPLES section on the index.md page.

janvanmansum commented on lib/src/test/java/nl/knaw/dans/dansschema/DdmVersionChangesTest.java:
```diff
@@ -0,0 +1,181 @@
+    }
+
+    @Test
+    public void v1_should_require_schemas_dans_knaw_nl_name_space_() throws Exception {
```
```suggestion
    public void v1_should_require_schemas_dans_knaw_nl_namespace() throws Exception {
```

janvanmansum commented on lib/src/test/java/nl/knaw/dans/dansschema/DdmVersionChangesTest.java:
```diff
@@ -0,0 +1,181 @@
+    }
+
+    @Test
+    public void v1_should_allow_additional_xml_in_root() throws Exception {
```
The test name does not explain well what is tested. What _actually_ seems to be tested is that an empty ddm is not valid. A good additional test would be to have a DDM with the minimally required content and show that is passes.

janvanmansum commented on lib/src/test/java/nl/knaw/dans/dansschema/DdmVersionChangesTest.java:
```diff
@@ -0,0 +1,181 @@
+    }
+
+    @Test
+    public void v2_should_not_allow_additional_xml_in_root() throws Exception {
```
Again, the name doesn't match what is tested. Even less, I should say, as additional-xml is not even mentioned in the error message.

janvanmansum commented on lib/src/test/java/nl/knaw/dans/dansschema/DdmVersionChangesTest.java:
```diff
@@ -0,0 +1,181 @@
+    }
+
+    @Test
+    public void v2_should_require_personal_data() throws Exception {
```
This is what all tests should look like! It says what is does and does what is says.

janvanmansum reviewed (changes requested):
Have another look at this. See my inline comments.

jo-pol committed:
Update lib/src/test/java/nl/knaw/dans/dansschema/DdmVersionChangesTest.java

Co-authored-by: Jan van Mansum <jan.van.mansum@dans.knaw.nl>

jo-pol committed:
v0 bag -> v1 bag

jo-pol committed:
merged readme into docs/index

jo-pol committed:
test names

jo-pol committed:
language code as in review request for related PR

https://github.com/DANS-KNAW/dd-dans-sword2-examples/pull/8#discussion_r1181522806

jo-pol committed:
Merge remote-tracking branch 'blessed/master' into DD-1229-schema-tests

janvanmansum reviewed (changes requested):
`dansschema` -> `schema`

janvanmansum commented on lib/src/test/java/nl/knaw/dans/dansschema/DdmVersionChangesTest.java:
```diff
@@ -0,0 +1,181 @@
+    }
+
+    @Test
+    public void v1_should_require_schemas_dans_knaw_nl_namespace() throws Exception {
```
```suggestion
    public void v1_should_require_easy_dans_knaw_nl_namespace() throws Exception {
```

janvanmansum commented on lib/src/test/java/nl/knaw/dans/dansschema/DdmVersionChangesTest.java:
```diff
@@ -0,0 +1,181 @@
+    }
+
+    @Test
+    public void v2_should_require_easy_dans_knaw_nl_name_space() throws Exception {
```
```suggestion
    public void v2_should_require_schemas_dans_knaw_nl_name_space() throws Exception {
```

janvanmansum commented on lib/src/test/java/nl/knaw/dans/dansschema/DdmVersionChangesTest.java:
```diff
@@ -0,0 +1,181 @@
+        assertThat(result).hasSize(1);
+        assertThat(result.get(0))
+            .hasMessageContaining("The content of element 'ddm:DDM' is not complete. One of")
+            .hasMessageContaining(":profile, ")
```
remove non-essentials (profile, dcmMetadata) from expectation

janvanmansum commented on lib/src/test/java/nl/knaw/dans/dansschema/DdmVersionChangesTest.java:
```diff
@@ -0,0 +1,181 @@
+    }
+
+    @Test
+    public void v1_should_not_suggest_additional_xml_for_root() throws Exception {
```
```suggestion
    public void v2_should_not_suggest_additional_xml_for_root() throws Exception {
```

janvanmansum commented on lib/src/test/java/nl/knaw/dans/dansschema/DdmVersionChangesTest.java:
```diff
@@ -0,0 +1,181 @@
+        assertThat(result.get(0))
+            .hasMessageContaining("The content of element 'ddm:DDM' is not complete. One of")
+            .hasMessageContaining(":profile, ")
+            .hasMessageContaining(":dcmiMetadata}' is expected.");
```
Test that message does NOT contain "additional-xml"

janvanmansum commented on lib/src/test/java/nl/knaw/dans/dansschema/DdmVersionChangesTest.java:
```diff
@@ -0,0 +1,181 @@
+        + "        <ddm:language encodingScheme='ISO639-2' code='ka'>Groenlands</ddm:language>"
+        + "        <ddm:language encodingScheme='ISO639-3' code='ba'>Baskisch</ddm:language>";
+
+    private static String simpleXml(String rootNameSpace, int nrOfTitles, String lastProfileElement, String dcmiElements) {
```
Replace with DdmBuilder helper class:

```
ddm = new DdmBuilder().withMultiTitle()....build().
```

janvanmansum reviewed (commented):
Remove readmes in test/resources

jo-pol committed:
Update lib/src/test/java/nl/knaw/dans/dansschema/DdmVersionChangesTest.java

Co-authored-by: Jan van Mansum <jan.van.mansum@dans.knaw.nl>

jo-pol committed:
Update lib/src/test/java/nl/knaw/dans/dansschema/DdmVersionChangesTest.java

Co-authored-by: Jan van Mansum <jan.van.mansum@dans.knaw.nl>

jo-pol committed:
Update lib/src/test/java/nl/knaw/dans/dansschema/DdmVersionChangesTest.java

Co-authored-by: Jan van Mansum <jan.van.mansum@dans.knaw.nl>

jo-pol committed:
essential expectations

jo-pol committed:
package change

jo-pol committed:
dropped readme's

jo-pol committed:
renames: clarify how xml files are found per xsd

jo-pol committed:
ddm builder

jo-pol committed:
rename of test methods

jo-pol committed:
typo

jo-pol committed:
license header

jo-pol committed:
rename test classes

jo-pol committed:
irregular false negatives

jo-pol committed:
ignore xml:lang, comments, doc

jo-pol committed:
moved matching dirs remark, typo

jo-pol committed:
Merge remote-tracking branch 'blessed/master' into DD-1229-schema-tests
